src/dbnode/storage/series/buffer.go

```diff
@@ -473,119 +440,346 @@ func (b *dbBuffer) FetchBlocksMetadata(
+	return nil, false
 }
 
-type inOrderEncoder struct {
-	encoder     encoding.Encoder
-	lastWriteAt time.Time
+func (b *dbBuffer) bucketVersionsAtCreate(
+	t time.Time,
+) *BufferBucketVersions {
+	if buckets, exists := b.bucketVersionsAt(t); exists {
+		return buckets
+	}
+
+	buckets := b.bucketVersionsPool.Get()
+	buckets.resetTo(t, b.opts, b.bucketPool)
+	b.bucketsMap[xtime.ToUnixNano(t)] = buckets
+	b.inOrderBlockStartsAdd(t)
+
+	return buckets
 }
 
-func (b *dbBufferBucket) resetTo(
-	start time.Time,
-) {
-	// Close the old context if we're resetting for use
-	b.finalize()
+func (b *dbBuffer) putBucketVersionsInCache(newBuckets *BufferBucketVersions) {
+	replaceIdx := bucketsCacheSize - 1
+	for i, buckets := range b.bucketVersionsCache {
+		// Check if we have the same pointer in cache.
+		if buckets == newBuckets {
+			replaceIdx = i
+		}
+	}
 
-	bopts := b.opts.DatabaseBlockOptions()
-	encoder := bopts.EncoderPool().Get()
-	encoder.Reset(start, bopts.DatabaseBlockAllocSize())
+	for i := replaceIdx; i > 0; i-- {
+		b.bucketVersionsCache[i] = b.bucketVersionsCache[i-1]
+	}
+
+	b.bucketVersionsCache[0] = newBuckets
+}
+
+func (b *dbBuffer) removeBucketVersionsInCache(oldBuckets *BufferBucketVersions) {
+	nilIdx := -1
+	for i, buckets := range b.bucketVersionsCache {
+		if buckets == oldBuckets {
+			nilIdx = i
+		}
+	}
+	if nilIdx == -1 {
+		return
+	}
+
+	for i := nilIdx; i < bucketsCacheSize-1; i++ {
+		b.bucketVersionsCache[i] = b.bucketVersionsCache[i+1]
+	}
+
+	b.bucketVersionsCache[bucketsCacheSize-1] = nil
+}
+
+func (b *dbBuffer) removeBucketVersionsAt(blockStart time.Time) {
+	buckets, exists := b.bucketVersionsAt(blockStart)
+	if !exists {
+		return
+	}
+	delete(b.bucketsMap, xtime.ToUnixNano(blockStart))
+	b.removeBucketVersionsInCache(buckets)
+	b.inOrderBlockStartsRemove(blockStart)
+	// nil out pointers.
+	buckets.resetTo(timeZero, nil, nil)
+	b.bucketVersionsPool.Put(buckets)
+}
+
+func (b *dbBuffer) inOrderBlockStartsAdd(newTime time.Time) {
+	starts := b.inOrderBlockStarts
+	idx := len(starts)
+	// There shouldn't be that many starts here, so just linear search through.
+	for i, t := range starts {
+		if t.After(newTime) {
+			idx = i
+			break
+		}
+	}
+	// Insert new time without allocating new slice.
+	b.inOrderBlockStarts = append(starts, timeZero)
+	// Update to new slice
+	starts = b.inOrderBlockStarts
+	copy(starts[idx+1:], starts[idx:])
+	starts[idx] = newTime
+}
 
+func (b *dbBuffer) inOrderBlockStartsRemove(removeTime time.Time) {
+	starts := b.inOrderBlockStarts
+	// There shouldn't be that many starts here, so just linear search through.
+	for i, t := range starts {
+		if t.Equal(removeTime) {
+			b.inOrderBlockStarts = append(starts[:i], starts[i+1:]...)
+			return
+		}
+	}
+}
+
+// BufferBucketVersions is a container for different versions (from
+// different flushes to disk) of buffer buckets in the database.
+type BufferBucketVersions struct {
+	buckets           []*BufferBucket
+	start             time.Time
```

Comment:
what is this start time for? 

each BufferBucket struct has a start time as well.

how do you decide which buckets go into which version?